pkg/controller/dns/letsencrypt.go

```diff
@@ -0,0 +1,295 @@
+
+	conf.CADirURL = u.URL
+
+	client, err := lego.NewClient(conf)
+	if err != nil {
+		return err
+	}
+
+	reg, err := client.Registration.Register(registration.RegisterOptions{TermsOfServiceAgreed: true})
+	if err != nil {
+		return err
+	}
+	u.Registration = reg
+
+	logrus.Infof("registered LE User: %s", u.Name)
+
+	return nil
+
+}
+
+func (u *LEUser) ToSecret() (*corev1.Secret, error) {
+	if u.Registration == nil {
+		return nil, fmt.Errorf("not saving LE User: missing registration")
+	}
+	x509Encoded, _ := x509.MarshalECPrivateKey(u.Key.(*ecdsa.PrivateKey))
+	pemEncoded := pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: x509Encoded})
+
+	reg, err := json.Marshal(u.Registration)
+	if err != nil {
+		return nil, err
+	}
+
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      system.LESecretName,
+			Namespace: system.Namespace,
+			Labels: map[string]string{
+				labels.AcornManaged: "true",
+			},
+		},
+		Data: map[string][]byte{
+			"name":         []byte(u.Name),
+			"email":        []byte(u.Email),
+			"privateKey":   pemEncoded,
+			"registration": reg,
+			"url":          []byte(u.URL),
+		},
+	}, nil
+}
+
+func (u *LEUser) GenerateWildcardCert(dnsendpoint, domain, token string) (*certificate.Resource, error) {
+	if u.Registration == nil {
+		return nil, fmt.Errorf("not generating LE cert: missing registration")
+	}
+
+	if domain == "" || token == "" {
+		return nil, fmt.Errorf("not generating LE cert: missing domain or token")
+	}
+
+	conf := lego.NewConfig(u)
+
+	conf.CADirURL = u.URL
+
+	client, err := lego.NewClient(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	dnsProvider := NewDNSProvider(dnsendpoint, domain, token)
+
+	if err := client.Challenge.SetDNS01Provider(dnsProvider); err != nil {
+		return nil, err
+	}
+
+	request := certificate.ObtainRequest{
+		Domains: []string{fmt.Sprintf("*.%s", strings.TrimPrefix(domain, "."))},
+		Bundle:  true,
+	}
+
+	return client.Certificate.Obtain(request)
+}
+
+func EnsureLEUser(ctx context.Context, client kclient.Client, domain string) (*LEUser, error) {
+
+	targetUsername := strings.TrimPrefix(domain, ".") // leading dot is an issue especially for email addresses
+
+	leAccountSecret := &corev1.Secret{}
+	err := client.Get(ctx, router.Key(system.Namespace, system.LESecretName), leAccountSecret)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	// Existing LE User Secret found
+	if err == nil {
+		leUser, err := FromSecret(leAccountSecret)
+		if err != nil {
+			return nil, err
+		}
+
+		// Domain changed, recreate the secret
+		if !strings.Contains(leUser.Name, targetUsername) {
+			logrus.Infof("deleting LE secret for domain %s", domain)
+			if err := client.Delete(ctx, leAccountSecret); err != nil {
+				return nil, err
+			}
+		} else {
+			return leUser, nil
+		}
+	}
+
+	// Create and Register Let's Encrypt User
+	leUser := &LEUser{
+		Name:  targetUsername,
+		Email: fmt.Sprintf("%s@on-acorn.io", targetUsername),
```

Comment:
We are going to need to prompt the user during install for an email address. We can't use a generic one, that got us in trouble with Rio when we did that.